src/CallStack.ts

```diff
@@ -353,33 +362,44 @@ export namespace CallStack {
 			return copy;
 		}
 
-		get jsonToken(){
-			let threadJObj: any = {};
+		public WriteJson(writer: SimpleJson.Writer) {
+			writer.WriteObjectStart();
 
-			let jThreadCallstack: any[] = [];
+			// callstack
+			writer.WritePropertyStart('callstack');
+			writer.WriteArrayStart();
 			for (let el of this.callstack) {
-				let jObj: any = {};
-				if (!el.currentPointer.isNull) {
+				writer.WriteObjectStart();
+				if(!el.currentPointer.isNull) {
 					if (el.currentPointer.container === null) { return throwNullException('el.currentPointer.container'); }
-					jObj['cPath'] = el.currentPointer.container.path.componentsString;
-					jObj['idx'] = el.currentPointer.index;
+					writer.WriteProperty('cPath', el.currentPointer.container.path.componentsString);
+					writer.WriteIntProperty('idx', el.currentPointer.index);
 				}
-				jObj['exp'] = el.inExpressionEvaluation;
-				jObj['type'] = el.type;
-				jObj['temp'] = JsonSerialisation.DictionaryRuntimeObjsToJObject(el.temporaryVariables);
-				jThreadCallstack.push(jObj);
+
+				writer.WriteProperty('exp', el.inExpressionEvaluation);
+				writer.WriteIntProperty('type', el.type);
+
+				if (el.temporaryVariables.size > 0) {
+					writer.WritePropertyStart('temp');
+					JsonSerialisation.WriteDictionaryRuntimeObjs(writer, el.temporaryVariables);
+					writer.WritePropertyEnd();
+				}
+
+				writer.WriteObjectEnd();
 			}
+			writer.WriteArrayEnd();
+			writer.WritePropertyEnd();
 
-			threadJObj['callstack'] = jThreadCallstack;
-			threadJObj['threadIndex'] = this.threadIndex;
+			// threadIndex
+			writer.WriteIntProperty('threadIndex', this.threadIndex);
 
 			if (!this.previousPointer.isNull) {
 				let resolvedPointer = this.previousPointer.Resolve();
 				if (resolvedPointer === null) { return throwNullException('this.previousPointer.Resolve()'); }
```

Comment:
Probably a silly question but I'm a bit rusty, remind me why we have these extra lines? Is it to please the typescript god?

Reply:
No silly questions, I actually had to be reminded myself.

It's to please the typescript god indeed and follow the behaviour of C#. Typescript will complain if we try to access a variable which might be null, but is smart enough to detect paths where it can't be null since a check was performed. We have these lines in many of placed, they'll raise a "NullPointerException" like C#.

As I got back into it, I took notes of all the strange things. We really need to work on this  CONTRIBUTING document!

Reply:
What I find strange is that there wasn't a reason to use `previousPointer` or `resolvedPointer` previously, so… why was it a problem if they were null ?
Anyway, they make sense now!

Reply:
> What I find strange is that there wasn't a reason to use previousPointer or resolvedPointer previously, so… why was it a problem if they were null ?

I'm not following, what do you mean?

Reply:
Nevermind, I misread something in the C# diff which threw me off. It all makes sense now 👍